src/Open3D/GUI/FileDialog.cpp

```diff
@@ -0,0 +1,411 @@
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+// ----------------------------------------------------------------------------
+
+#include "FileDialog.h"
+
+#include "Button.h"
+#include "Combobox.h"
+#include "Label.h"
+#include "Layout.h"
+#include "ListView.h"
+#include "TextEdit.h"
+#include "Theme.h"
+#include "Util.h"
+
+#include "Open3D/Utility/Console.h"
+#include "Open3D/Utility/Helper.h"
+#include "Open3D/Utility/FileSystem.h"
+
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+
+// macOS sorts directories in with the files
+// Windows and Linux (GTK) sort directories first.
+#ifdef __APPLE__
+#define INLINE_DIRS 1
+#else
+#define INLINE_DIRS 0
+#endif // __APPLE__
+
+namespace open3d {
+namespace gui {
+
+class DirEntry {
+public:
+    enum Type { DIR, FILE };
+
+    DirEntry(const std::string& name, Type type) {
+        type_ = type;
+        name_ = name;
+        if (type == DIR) {
+            display_ = std::string("[ ] ") + name_ + "/";
+        } else {
+            display_ = std::string("    ") + name_;
+        }
+    }
+
+    Type GetType() const { return type_; }
+    const std::string& GetName() const { return name_; }
+    const std::string& GetDisplayText() const { return display_; }
+
+    bool operator==(const DirEntry& rhs) const {
+        return (type_ == rhs.type_ && name_ == rhs.name_);
+    }
+    bool operator!=(const DirEntry& rhs) const { return !this->operator==(rhs); }
+
+    bool operator<(const DirEntry& rhs) const {
+#if INLINE_DIRS
+        // Sort directories by name; if the OS allows directories and files
+        // to have the same name, put directories first.
+        if (name_ == rhs.name_) {
+            if (type_ == rhs.type_) {
+                return false;
+            } else {
+                return (type_ == DIR);
+            }
+        } else {
+            return (name_ < rhs.name_);
+        }
+#else
+        // Sort directories first, then files.
+        // Within each category sort by name.
+        if (type_ == rhs.type_) {
+            return (name_ < rhs.name_);
+        } else {
+            return (type_ == DIR);
+        }
+#endif // INLINE_DIRS
+    }
+
+private:
+    Type type_;
+    std::string name_;
+    std::string display_;
+};
+
+struct FileDialog::Impl {
+    Type type;
+    std::vector<DirEntry> entries;
+    std::shared_ptr<TextEdit> filename;
+    std::shared_ptr<Combobox> dirtree;
+    std::shared_ptr<ListView> filelist;
+    std::shared_ptr<Combobox> filter;
+    std::unordered_map<int, std::unordered_set<std::string>> filterIdx2filter;
+    std::shared_ptr<Horiz> filterRow;
+    std::shared_ptr<Button> ok;
+    std::shared_ptr<Button> cancel;
+    std::function<void()> onCancel;
+    std::function<void(const char *)> onDone;
+
+    const DirEntry& GetSelectedEntry() {
+        static DirEntry gBogus("", DirEntry::Type::FILE);
+
+        int idx = this->filelist->GetSelectedIndex();
```

Comment:
I use this-> for clarity when referencing member variables that aren't mMember, _member, or member_. I discovered that the Impl was better suited to some things than the FileDialog class was, and since I haven't been postfixing underscores on structs, I prefixed it with this-> for clarity.